group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -682,15 +682,43 @@ CompletableFuture<AlterShareGroupOffsetsResponseData> persisterInitialize(
                         handlePersisterInitializeResponse(request.groupTopicPartitionData().groupId(), result, new ShareGroupHeartbeatResponseData());
                         return response;
                     } else {
-                        //TODO build new AlterShareGroupOffsetsResponseData for error response
-                        return response;
+                        return buildErrorResponse(response, result);
                     }
                 } else {
                     return buildErrorResponse(request, response, exp);
                 }
 
             });
     }
+    
+    private AlterShareGroupOffsetsResponseData buildErrorResponse(AlterShareGroupOffsetsResponseData response, InitializeShareGroupStateResult result) {
+        AlterShareGroupOffsetsResponseData data = new AlterShareGroupOffsetsResponseData();
+        data.setResponses(
+            new AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponseTopicCollection(response.responses().stream()
+                .map(topic -> {
+                    AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponseTopic topicData = new AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponseTopic()
+                        .setTopicName(topic.topicName());
+                    topic.partitions().forEach(partition -> {
+                        AlterShareGroupOffsetsResponseData.AlterShareGroupOffsetsResponsePartition partitionData;
+                        PartitionErrorData error = result.getErrors().get(topic.topicId()).get(partition.partitionIndex());
```

Comment:
`result.getErrors()` will create a new collection by grouping data. Perhaps it should be moved out of loop.

Reply:
Furthermore, how do we ensure that the topic from `AlterShareGroupOffsetsResponseData` must exist in `InitializeShareGroupStateResult`?

If users attempt  to alter the nonexistent partitions within an existing topic, the `AlterShareGroupOffsetsResponseData` will contain the topic but `InitializeShareGroupStateResult` will not. Is this correct?